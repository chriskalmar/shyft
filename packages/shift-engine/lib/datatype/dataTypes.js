
import DataType from './DataType';
import DataTypeUser from './DataTypeUser';

import casual from 'casual';
import _ from 'lodash';


export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(-2^10, 2^10),
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  mock: /* istanbul ignore next */
    () => casual.double(-2^10, 2^10),
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  mock: /* istanbul ignore next */
    () => casual.boolean,
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  mock: /* istanbul ignore next */
    () => casual.title,
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  mock: /* istanbul ignore next */
    randomJson,
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


/* istanbul ignore next */
function randomJson() {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase( casual.words(2) )
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer

    ret[ key ] = value
  })

  return ret
}
