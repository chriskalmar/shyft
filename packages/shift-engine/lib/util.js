import casual from 'casual';
import _ from 'lodash';



export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])



export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap }
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}


export const isString = (str) => (typeof str === 'string')


export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work')
  }

  return { ...first, ...second }
}



export const mapOverProperties = (object, iteratee) => {

  passOrThrow(
    isMap(object),
    () => 'Provided object is not a map'
  )

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function'
  )


  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[ key ], key)
  })
}



export const sortDataByKeys = (keys, data, keyProperty='id') => {
  const map = {}
  const result = []
  const order = {}

  if (!keys || ( isArray(keys) && keys.length === 0)) {
    return []
  }

  if (!data || ( isArray(data) && data.length === 0)) {
    return keys.map(() => null)
  }


  for (let d = 0; d < data.length; d++) {
    const row = data[ d ]
    const id = row[ keyProperty ]
    const found = map[id]

    if (!found) {
      map[id] = row
    }
    else if (found instanceof Array) {
      found.push(row)
    }
    else {
      map[id] = [ found, row ];
    }
  }


  for (let k = 0; k < keys.length; k++) {
    const key = keys[ k ]
    const found = map[ key ];

    if (typeof found === 'undefined') {
      result.push(null)
    }
    else if (found instanceof Array) {
      let idx = order[ key ];

      if (typeof idx === 'undefined') {
        idx = 0
      }
      else {
        idx++
      }

      order[ key ] = idx;

      result.push(found[ idx ]);
    }
    else {
      result.push(found);
    }
  }

  return result;
};



export const processCursor = (entity, cursor, orderBy, reverse) => {

  const $LT = reverse ? '$gt' : '$lt'
  const $GT = reverse ? '$lt' : '$gt'
  const $LTE = reverse ? '$gte' : '$lte'
  const $GTE = reverse ? '$lte' : '$gte'

  const where = {}

  if (cursor) {
    passOrThrow(
      isArray(cursor[ entity.name ]),
      () => 'Incompatible cursor for this entity'
    )

    passOrThrow(
      isArray(orderBy),
      () => 'orderBy needs to be an array of order definitions'
    )

    const orderList = []

    const orderMap = {}
    orderBy.map(({attribute, direction}) => {

      passOrThrow(
        attribute && direction,
        () => 'orderBy needs to be an array of attributes and respective sort order'
      )

      orderMap[ attribute ] = direction
    })

    const primaryAttribute = entity.getPrimaryAttribute()
    const attributes = entity.getAttributes()

    let foundUniqueAttribute = false

    cursor[ entity.name ].map(filter => {
      if (filter.length !== 2) {
        throw new Error('Cursor malformed')
      }

      const attributeName = filter[0]

      passOrThrow(
        attributes[ attributeName ],
        () => `Unknown attribute '${attributeName}' used in cursor`
      )

      if (attributeName !== primaryAttribute.name) {
        passOrThrow(
          orderMap[ attributeName ],
          () => `Cursor works only on sorted attributes (check: '${attributeName}')`
        )
      }

      const attribute = attributes[ attributeName ]

      // limit where clause to the first attribute which is defined as unique
      if (!foundUniqueAttribute) {
        orderList.push(attributeName)
      }

      if (attribute.isUnique) {
        foundUniqueAttribute = true
      }

    })

    passOrThrow(
      foundUniqueAttribute,
      () => 'Cursor needs to have at least one attribute defined as unique'
    )


    // if more than 2 attributes are used for the cursor take only the first and the last (primary key)
    if (orderList.length > 2) {
      orderList.splice(1, (orderList.length - 2) )
    }


    // simple filter for single attributes
    if (orderList.length === 1) {

      const attributeName = orderList[0]
      const value = cursor[ entity.name ][0][1]

      if (orderMap[ attributeName ] === 'DESC') {
        where[ attributeName ] = {
          [ $LT ]: value
        }
      }
      else {
        where[ attributeName ] = {
          [ $GT ]: value
        }
      }

    }
    else {
      where.$not = {}

      cursor[ entity.name ].map(filter => {
        const attributeName = filter[0]
        const value = filter[1]

        // ignore attributes that obsolete due to prior unique attribute
        if (orderList.indexOf(attributeName) === -1) {
          return
        }

        const attribute = attributes[ attributeName ]

        if (attribute.isUnique) {
          if (orderMap[ attributeName ] === 'DESC') {
            where.$not[ attributeName ] = {
              [ $GTE ]: value
            }
          }
          else {
            where.$not[ attributeName ] = {
              [ $LTE ]: value
            }
          }
        }
        else {
          where.$not[ attributeName ] = value

          if (orderMap[ attributeName ] === 'DESC') {
            where[ attributeName ] = {
              [ $LTE ]: value
            }
          }
          else {
            where[ attributeName ] = {
              [ $GTE ]: value
            }
          }
        }

      })
    }
  }

  return where
}


export const processCursors = (entity, args) => {

  const {
    after,
    before,
  } = args;

  const where = {
    $and: [
      processCursor(entity, after, args.orderBy),
      processCursor(entity, before, args.orderBy, true),
    ]
  }

  return where
}



export const reverseString = (str) => {
  return str
    .split('')
    .reverse()
    .join('');
}



export const randomJson = () => {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase(casual.words(2))
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer()

    ret[key] = value
  })

  return ret
}
