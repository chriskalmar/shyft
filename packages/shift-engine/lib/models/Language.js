
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';


export const Language = new Entity({
  name: 'Language',
  description: 'A language',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'name' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'isoCode' ],
    }),
  ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      required: true
    },
  }

})

