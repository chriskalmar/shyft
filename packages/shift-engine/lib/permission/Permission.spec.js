
import Permission, {
  isPermission,
  isPermissionsArray,
  findInvalidPermissionAttributes,
  findMissingPermissionAttributes,
  generatePermissionDescription,
  checkPermissionSimple,
  buildUserAttributesPermissionFilter,
  buildStatesPermissionFilter,
  buildPermissionFilterSingle,
  buildPermissionFilter,
  processEntityPermissions,
} from './Permission';
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { passOrThrow } from '../util';
import { Language } from '../models/Language';


describe('Permission', () => {

  it('should create permission objects', () => {

    new Permission().everyone()
    new Permission().authenticated()
    new Permission().role('manager')
    new Permission().lookup(Language, {
      createdBy: 'someAttribute',
    })
    new Permission().value('someAttribute', 123)

    expect(String(new Permission())).toBe('Permission Object')
  })


  it('should be combinable', () => {

    const permission = new Permission()
      .role('manager')
      .lookup(Language, {
        createdBy: 'someAttribute',
      })
      .value('someAttribute', 123)
      .role('admin')
      .lookup(Language, {
        country: 'any',
      })
      .value('someAttribute', 987)


    expect(permission.roles).toEqual([ 'manager', 'admin' ]);
    expect(permission.lookups).toEqual([
      {
        entity: Language,
        lookupMap: {
          createdBy: 'someAttribute',
        }
      },
      {
        entity: Language,
        lookupMap: {
          country: 'any',
        }
      }
    ]);
    expect(permission.values).toEqual([
      {
        attributeName: 'someAttribute',
        value: 123
      },
      {
        attributeName: 'someAttribute',
        value: 987
      }
    ]);

  })


  it('should reject incompatible permission types', () => {

    function fn1() {
      new Permission()
        .role('manager')
        .everyone()
    }

    expect(fn1).toThrowErrorMatchingSnapshot();


    function fn2() {
      new Permission()
        .userAttribute('author')
        .everyone()
    }

    expect(fn2).toThrowErrorMatchingSnapshot();


    function fn3() {
      new Permission()
        .everyone()
        .authenticated()
    }

    expect(fn3).toThrowErrorMatchingSnapshot();


  })



  describe('role permissions', () => {

    it('should reject if role name is missing', () => {

      function fn1() {
        new Permission()
          .role()
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })


    it('should reject if duplicate roles are provided', () => {

      function fn1() {
        new Permission()
          .role('manager')
          .role('manager')
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })

  })


  describe('userAttribute permissions', () => {

    it('should reject if attribute name is missing', () => {

      function fn1() {
        new Permission()
          .userAttribute()
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })

    it('should reject if duplicate attribute names are provided', () => {

      function fn1() {
        new Permission()
          .userAttribute('profile')
          .userAttribute('profile')
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })

  })


  describe('lookup permissions', () => {

    it('should reject if entity is missing', () => {

      function fn1() {
        new Permission()
          .lookup()
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })


    it('should reject if lookupMap is missing', () => {

      function fn1() {
        new Permission()
          .lookup(Language)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })

  })


  describe('value permissions', () => {

    it('should reject if attribute name is missing', () => {

      function fn1() {
        new Permission()
          .value()
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })


    it('should reject if value is missing', () => {

      function fn1() {
        new Permission()
          .value('someAttribute')
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

    })

  })



  describe('isPermission', () => {


    it('should recognize objects of type Permission', () => {

      const permission = new Permission()

      function fn() {
        passOrThrow(
          isPermission(permission),
          () => 'This error will never happen'
        )
      }

      expect(fn).not.toThrow()

    })


    it('should recognize non-Permission objects', () => {

      function fn() {
        passOrThrow(
          isPermission({}) ||
          isPermission(function test() {}) ||
          isPermission(Error),
          () => 'Not a Permission object'
        )
      }


      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })

  describe('isPermissionsArray', () => {


    it('should recognize a list of objects of type Permission', () => {

      const permission = new Permission()

      function fn() {
        passOrThrow(
          isPermissionsArray([permission]) &&
          isPermissionsArray([permission, permission]) &&
          isPermissionsArray([permission, permission, permission]),
          () => 'This error will never happen'
        )
      }

      expect(fn).not.toThrow()

    })


    it('should recognize non-Permission objects', () => {

      const permission = new Permission()

      function fn() {
        passOrThrow(
          isPermissionsArray() ||
          isPermissionsArray(null) ||
          isPermissionsArray([]) ||
          isPermissionsArray({}) ||
          isPermissionsArray(function test() {}) ||
          isPermissionsArray(Error) ||
          isPermissionsArray([{}]) ||
          isPermissionsArray([function test() {}]) ||
          isPermissionsArray([Error]) ||
          isPermissionsArray([permission, null]) ||
          isPermissionsArray([permission, {}, permission]),
          () => 'Not a Permissions array'
        )
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })


  describe('permission attributes', () => {

    const City = new Entity({
      name: 'City',
      description: 'A city',
      attributes: {
        cityName: {
          type: DataTypeString,
          description: 'City name'
        },
      }
    })

    const User = new Entity({
      name: 'User',
      description: 'A user',
      isUserEntity: true,
      attributes: {
        loginName: {
          type: DataTypeString,
          description: 'Login name'
        },
        city: {
          type: City,
          description: 'User\'s city'
        }
      }
    })


    it('should reject if userAttribute is not a reference to the user entity', () => {

      function fn1() {
        const permission = new Permission()
          .userAttribute('any')

        findInvalidPermissionAttributes(permission, City)
      }

      function fn2() {
        const permission = new Permission()
          .userAttribute('cityName')

        findInvalidPermissionAttributes(permission, City)
      }

      function fn3() {
        const permission = new Permission()
          .userAttribute('city')

        findInvalidPermissionAttributes(permission, User)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();

    })


    it('should find missing permission attributes', () => {

      {
        const permission = new Permission()
          .userAttribute('wrong')

        const missing = findMissingPermissionAttributes(permission, City)

        expect(missing).toEqual('wrong')
      }

      {
        const permission = new Permission()
          .lookup(User, {
            id: 'wrong'
          })

        const missing = findMissingPermissionAttributes(permission, City)

        expect(missing).toEqual('User.wrong')
      }

      {
        const permission = new Permission()
          .lookup(User, {
            hello: 'city'
          })

        const missing = findMissingPermissionAttributes(permission, City)

        expect(missing).toEqual('hello')
      }

      {
        const permission = new Permission()
          .value('wrong', 123)

        const missing = findMissingPermissionAttributes(permission, City)

        expect(missing).toEqual('wrong')
      }

    })


    it('should accept correctly defined permission attributes', () => {

      const permission = new Permission()
        .userAttribute('id')
        .lookup(User, {
          id: 'city'
        })
        // https://en.wikipedia.org/wiki/Taumatawhakatangihangakoauauotamateaturipukakapikimaungahoronukupokaiwhenuakitanatahu
        .value('cityName', 'Taumatawhakatangihangakoauauotamateaturipukakapikimaungahoronukupokaiwhenuakitanatahu')

      const missing = findMissingPermissionAttributes(permission, City)

      expect(missing).toBe(false)
    })


  })


  describe('permissions description', () => {

    it('should reject if non-Permission object is provided', () => {

      function fn() {
        generatePermissionDescription({foo: 'bar'})
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should generate a description based on defined permissions', () => {

      const tests = [
        [
          new Permission().everyone(),
          'everyone'
        ],
        [
          new Permission().authenticated(),
          'authenticated'
        ],
        [
          new Permission().role('manager'),
          'role manager'
        ],
        [
          new Permission().userAttribute('publisher'),
          'userAttributes publisher'
        ],
        [
          new Permission().lookup(Language, { createdBy: 'someAttribute' }),
          'lookup Language someAttribute'
        ],
        [
          new Permission().value('someAttribute', 123),
          'value someAttribute'
        ],
        [
          new Permission()
            .role('manager')
            .role('admin')
            .userAttribute('publisher')
            .userAttribute('organizer')
            .lookup(Language, { createdBy: 'someAttribute' })
            .lookup(Language, { updatedAt: 'anotherAttribute' })
            .lookup(Language, { source: 'lorem', mainContinent: 'ipsum' })
            .value('someAttribute', 123)
            .value('anotherAttribute', 'hello'),
          'mixed 1'
        ],
      ]

      tests.map(([ permission, testName ]) => {
        expect(generatePermissionDescription(permission)).toMatchSnapshot(testName)
      })

    })


    it('should return no description if permission object is empty', () => {

      const permission = new Permission()

      const result = generatePermissionDescription(permission)

      expect(result).toBe(false)
    })

  })


  describe('permissions check simple', () => {

    const userId = 123
    const userRoles = [ 'manager', 'reviewer' ]

    it('should reject if non-Permission object is provided', () => {

      function fn() {
        checkPermissionSimple({})
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should reject if user roles are not provided as an array', () => {

      function fn() {
        checkPermissionSimple(new Permission(), null, {bad: 'roles'})
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should always give access if permission mode is `everyone`', () => {

      expect(checkPermissionSimple(
        new Permission().everyone()
      )).toBe(true)
    })


    it('should give access to authenticated users if permission mode is `authenticated`', () => {

      expect(checkPermissionSimple(
        new Permission().authenticated(),
        userId,
      )).toBe(true)
    })


    it('should reject access for anonymous users if permission mode is `authenticated`', () => {

      expect(checkPermissionSimple(
        new Permission().authenticated(),
      )).toBe(false)
    })


    it('should give access to users with corresponding user roles on permission mode `role`', () => {

      expect(checkPermissionSimple(
        new Permission().role('reviewer'),
        userId,
        userRoles,
      )).toBe(true)
    })


    it('should reject access for users with different roles on permission mode `role`', () => {

      expect(checkPermissionSimple(
        new Permission().role('admin'),
        userId,
        userRoles,
      )).toBe(false)

      expect(checkPermissionSimple(
        new Permission().authenticated().role('admin'),
        userId,
        userRoles,
      )).toBe(false)
    })


    it('should default to give access if neither `everyone`, `authenticated` or `role` is defined', () => {

      expect(checkPermissionSimple(
        new Permission(),
        userId,
        userRoles,
      )).toBe(true)
    })

  })


  describe('build permission filter', () => {

    const userId = 123
    const userRoles = ['manager', 'reviewer']

    const someEntity = new Entity({
      name: 'SomeEntityName',
      description: 'Just some description',
      attributes: {
        someAttribute: {
          type: DataTypeString,
          description: 'Just some description',
        },
      },
      states: {
        open: 10,
        closed: 20,
        inTransfer: 40,
        onHold: 50,
      },
    })


    describe('userAttributes', () => {

      it('should reject if userId is not provided', () => {

        function fn() {
          const permission = new Permission()
            .userAttribute('author')
          buildUserAttributesPermissionFilter(permission)
        }

        expect(fn).toThrowErrorMatchingSnapshot();
      })


      it('should return undefined filters if permission type is not used', () => {

        const permission = new Permission()
          .state('completed')

        const filter = buildUserAttributesPermissionFilter(permission, userId)

        expect(filter).toBeUndefined()
      })


      it('should generate filters for single entries', () => {

        const permission = new Permission()
          .userAttribute('author')

        const filter = buildUserAttributesPermissionFilter(permission, userId)

        expect(filter).toMatchSnapshot();
      })


      it('should generate filters for multiple entries', () => {

        const permission = new Permission()
          .authenticated()
          .userAttribute('author')
          .userAttribute('reviewer')

        const filter = buildUserAttributesPermissionFilter(permission, userId)

        expect(filter).toMatchSnapshot();
      })

    })


    describe('states', () => {

      it('should reject if entity is not provided', () => {

        function fn() {
          const permission = new Permission()
            .state('completed')
          buildStatesPermissionFilter(permission)
        }

        expect(fn).toThrowErrorMatchingSnapshot();
      })


      it('should reject if invalid state is used', () => {

        function fn() {
          const permission = new Permission()
            .state('completed')
          buildStatesPermissionFilter(permission, someEntity)
        }

        expect(fn).toThrowErrorMatchingSnapshot();
      })


      it('should return undefined filters if permission type is not used', () => {

        const permission = new Permission()
          .userAttribute('reviewer')

        const filter = buildStatesPermissionFilter(permission, someEntity)

        expect(filter).toBeUndefined()
      })


      it('should generate filters for single entries', () => {

        const permission = new Permission()
          .state('open')

        const filter = buildStatesPermissionFilter(permission, someEntity)

        expect(filter).toMatchSnapshot();
      })


      it('should generate filters for multiple entries', () => {

        const permission = new Permission()
          .authenticated()
          .state('open')
          .state('inTransfer')

        const filter = buildStatesPermissionFilter(permission, someEntity)

        expect(filter).toMatchSnapshot();
      })

    })


    describe('all permission types', () => {

      it('should return undefined filters if no permission type is used', () => {

        const permission = new Permission()

        const filter = buildPermissionFilterSingle(permission, userId, userRoles, someEntity)

        expect(filter).toBeUndefined()
      })


      it('should generate filters for single permission types', () => {

        const permission = new Permission()
          .authenticated()
          .userAttribute('author')

        const filter = buildPermissionFilterSingle(permission, userId, userRoles, someEntity)

        expect(filter).toMatchSnapshot();
      })


      it('should generate filters for multiple permission types', () => {

        const permission = new Permission()
          .authenticated()
          .userAttribute('author')
          .state('open')
          .state('inTransfer')

        const filter = buildPermissionFilterSingle(permission, userId, userRoles, someEntity)

        expect(filter).toMatchSnapshot();
      })

    })


    describe('combine multiple permissions', () => {

      it('should return undefined filters if no permission type is used', () => {

        const permission = new Permission()

        const filter = buildPermissionFilter(permission, userId, userRoles, someEntity)

        expect(filter).toBeUndefined()
      })


      it('should return undefined filters if request does not pass simple permission checks', () => {

        const permission1 = new Permission()
          .authenticated()
          .state('open')
          .state('inTransfer')

        const filter1 = buildPermissionFilter(permission1, null, null, someEntity)

        expect(filter1).toBeUndefined()


        const permission2 = new Permission()
          .authenticated()
          .role('customer')
          .state('open')
          .state('inTransfer')

        const filter2 = buildPermissionFilter(permission2, userId, userRoles, someEntity)

        expect(filter2).toBeUndefined()


        const multiPermissions = [
          new Permission()
            .authenticated()
            .role('customer')
            .state('open')
            .state('inTransfer'),
          new Permission()
            .role('agent')
            .userAttribute('createdBy')
        ]

        const filter3 = buildPermissionFilter(multiPermissions, userId, userRoles, someEntity)

        expect(filter3).toBeUndefined()
      })


      it('should generate filters for single permissions', () => {

        const permission = new Permission()
          .authenticated()
          .userAttribute('author')
          .state('open')

        const filter = buildPermissionFilter(permission, userId, userRoles, someEntity)

        expect(filter).toMatchSnapshot();
      })


      it('should generate filters for multiple permissions', () => {

        const multiPermissions = [
          new Permission()
            .authenticated()
            .userAttribute('author')
            .state('open')
            .state('inTransfer'),
          new Permission()
            .role('reviewer')
            .userAttribute('createdBy')
        ]

        const filter = buildPermissionFilter(multiPermissions, userId, userRoles, someEntity)

        expect(filter).toMatchSnapshot();
      })


      it('should generate filters for eligible permissions only', () => {

        const multiPermissions = [
          new Permission()
            .authenticated()
            .role('customer')
            .state('open'),
          new Permission()
            .authenticated()
            .userAttribute('author')
            .state('open')
            .state('inTransfer'),
          new Permission()
            .role('reviewer')
            .userAttribute('createdBy'),
          new Permission()
            .role('admin')
            .userAttribute('author')
        ]

        const filter = buildPermissionFilter(multiPermissions, userId, userRoles, someEntity)

        expect(filter).toMatchSnapshot();
      })

    })

  })


  describe('processEntityPermissions', () => {

    const entity = new Entity({
      name: 'SomeEntityName',
      description: 'Just some description',
      attributes: {
        someAttribute: {
          type: DataTypeString,
          description: 'Some description',
        }
      }
    })


    it('should accept a correct permissions setup', () => {

      const permissions = {
        read: new Permission().value('someAttribute', 123),
        mutations: {
          update: new Permission().role('manager'),
        }
      }

      processEntityPermissions(entity, permissions)
    })


    it('should throw if provided with an invalid map of permissions', () => {

      const permissions = [ 'bad' ]

      function fn() {
        processEntityPermissions(entity, permissions)
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should throw if provided with an invalid map of mutation permissions', () => {

      const permissions = {
        mutations: [ 'bad' ]
      }

      function fn() {
        processEntityPermissions(entity, permissions)
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should throw if provided with an invalid permissions', () => {

      const permissions1 = {
        read: [ 'bad' ]
      }

      function fn1() {
        processEntityPermissions(entity, permissions1)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();


      const permissions2 = {
        find: [ 'bad' ]
      }

      function fn2() {
        processEntityPermissions(entity, permissions2)
      }

      expect(fn2).toThrowErrorMatchingSnapshot();


      const permissions3 = {
        mutations: {
          mutations: {
            create: [ 'bad' ]
          }
        }
      }

      function fn3() {
        processEntityPermissions(entity, permissions3)
      }

      expect(fn3).toThrowErrorMatchingSnapshot();

    })


    it('should throw if permissions have unknown attributes defined', () => {

      const permissions1 = {
        read: new Permission().userAttribute('notHere')
      }

      function fn1() {
        processEntityPermissions(entity, permissions1)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();


      const permissions2 = {
        find: new Permission().value('notHere', 123)
      }

      function fn2() {
        processEntityPermissions(entity, permissions2)
      }


      expect(fn2).toThrowErrorMatchingSnapshot();


      const permissions3 = {
        mutations: {
          update: new Permission().userAttribute('notHere')
        }
      }

      function fn3() {
        processEntityPermissions(entity, permissions3)
      }

      expect(fn3).toThrowErrorMatchingSnapshot();

    })


    it('should throw if permissions have invalid attributes defined', () => {

      const permissions = {
        read: new Permission().userAttribute('someAttribute')
      }

      function fn() {
        processEntityPermissions(entity, permissions)
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })


    it('should throw if permissions are assigned to unknown mutations', () => {

      const permissions = {
        mutations: {
          noSuchMutation: new Permission().userAttribute('someAttribute')
        }
      }

      function fn() {
        processEntityPermissions(entity, permissions)
      }

      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })

})
