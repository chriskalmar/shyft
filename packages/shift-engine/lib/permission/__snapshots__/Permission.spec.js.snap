// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permission isPermission should recognize non-Permission objects 1`] = `"Not a Permission object"`;

exports[`Permission lookup permissions should reject if entity is missing 1`] = `"Permission type 'lookup' expects an entity"`;

exports[`Permission lookup permissions should reject if lookupMap is missing 1`] = `"Permission type 'lookup' expects a lookupMap"`;

exports[`Permission userAttribute permissions should reject if attribute name is missing 1`] = `"Permission type 'userAttribute' expects an attribute name"`;

exports[`Permission userAttribute permissions should reject if duplicate attribute names are provided 1`] = `"Duplicate attribute name 'profile' for permission type 'userAttribute'"`;

exports[`Permission permission attributes should reject if userAttribute is not a reference to the user entity 1`] = `"Cannot use attribute 'any' in 'City.permissions' as 'userAttribute' as it is not a reference to the User entity"`;

exports[`Permission permission attributes should reject if userAttribute is not a reference to the user entity 2`] = `"Cannot use attribute 'cityName' in 'City.permissions' as 'userAttribute' as it is not a reference to the User entity"`;

exports[`Permission permission attributes should reject if userAttribute is not a reference to the user entity 3`] = `"Cannot use attribute 'city' in 'User.permissions' as 'userAttribute' as it is not a reference to the User entity"`;

exports[`Permission permissions check simple should reject if non-Permission object is provided 1`] = `"checkPermissionSimple needs a valid permission object"`;

exports[`Permission permissions check simple should reject if user roles are not provided as an array 1`] = `"checkPermissionSimple needs a valid list of assigned user roles"`;

exports[`Permission permissions description should reject if non-Permission object is provided 1`] = `"generatePermissionDescription needs a valid permission object"`;

exports[`Permission processEntityPermissions should throw if permissions are assigned to unknown mutations 1`] = `"Unknown mutation 'noSuchMutation' used for permissions in entity 'SomeEntityName'"`;

exports[`Permission processEntityPermissions should throw if permissions have invalid attributes defined 1`] = `"Cannot use attribute 'someAttribute' in 'SomeEntityName.permissions' as 'userAttribute' as it is not a reference to the User entity"`;

exports[`Permission processEntityPermissions should throw if permissions have unknown attributes defined 1`] = `"Cannot use attribute 'notHere' in 'SomeEntityName.permissions' for 'read' as it does not exist"`;

exports[`Permission processEntityPermissions should throw if permissions have unknown attributes defined 2`] = `"Cannot use attribute 'notHere' in 'SomeEntityName.permissions' for 'find' as it does not exist"`;

exports[`Permission processEntityPermissions should throw if permissions have unknown attributes defined 3`] = `"Cannot use attribute 'notHere' in 'SomeEntityName.permissions' for 'update' as it does not exist"`;

exports[`Permission processEntityPermissions should throw if provided with an invalid map of mutation permissions 1`] = `"Entity 'SomeEntityName' permissions definition for mutations needs to be a map of mutations and permissions"`;

exports[`Permission processEntityPermissions should throw if provided with an invalid map of permissions 1`] = `"Entity 'SomeEntityName' permissions definition needs to be an object"`;

exports[`Permission processEntityPermissions should throw if provided with an invalid permissions 1`] = `"Invalid 'read' permission definition for entity 'SomeEntityName'"`;

exports[`Permission processEntityPermissions should throw if provided with an invalid permissions 2`] = `"Invalid 'find' permission definition for entity 'SomeEntityName'"`;

exports[`Permission processEntityPermissions should throw if provided with an invalid permissions 3`] = `"Invalid mutation permission definition for entity 'SomeEntityName' at position '0'"`;

exports[`Permission role permissions should reject if duplicate roles are provided 1`] = `"Duplicate role 'manager' for permission type 'role'"`;

exports[`Permission role permissions should reject if role name is missing 1`] = `"Permission type 'role' expects an role name"`;

exports[`Permission should reject incompatible permission types 1`] = `"Permission type 'everyone' is incompatible with other types"`;

exports[`Permission should reject incompatible permission types 2`] = `"Permission type 'authenticated' is incompatible with other types"`;

exports[`Permission should reject incompatible permission types 3`] = `"Permission type 'authenticated' is incompatible with other types"`;

exports[`Permission value permissions should reject if attribute name is missing 1`] = `"Permission type 'value' expects an attribute name"`;

exports[`Permission value permissions should reject if value is missing 1`] = `"Permission type 'value' expects a value"`;
